### PR TITLE
[SPARK-25354][SQL] Reduce unneeded operation in nextKeyValue process of parquet vectorized record reader

### PR DESCRIPTION
**Pull request by SongYadong**
## What changes were proposed in this pull request?

this PR do following in VectorizedParquetRecordReader class:

1. In `nextKeyValue()`, remove calling to `resultBatch()`. Columnar batch initializition by `initBatch()` is moved to `nextBatch()`.

2. In `nextBatch()`, put `if (rowsReturned >= totalRowCount) return false;` before columnVectors resetting. When condition is true, columnVectors resetting need not execute. 

3. In `nextBatch()`, call checkEndOfRowGroup only when `rowsReturned == totalCountLoadedSoFar` is true instead of calling every time.

4. In `checkEndOfRowGroup()`, save columns of requestedSchema for next use instead of reading every time. 

From analysis of JMC tool, we found that frequencies of calling to `nextKeyValue()` is very high, performance gain may be got from these changes.

## How was this patch tested?

1. Existing Unit Tests
2. A test of running 4885 spark applications.

### CODE REVIEW

**Comment by SongYadong** [comment]
In my test, total executing time of 4885 spark applications:
origin: 447213 seconds
after : 444584 seconds
time saved : about 0.6%

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java [hunk 1]
```diff
@@ -154,8 +159,6 @@ public void close() throws IOException {
 
   @Override
   public boolean nextKeyValue() throws IOException {
-    resultBatch();
```
Comment (kiszk) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: I like this direction to reduce overhead. I confirmed that the current code calls `initBatch()` before calling `nextKeyValue()`, too.  
My question is that is calling sequence is mandatory or optional. In the future, I wonder a developer may call this method thru `hasNext()` before calling `initBatch()`.

How about moving `if-statement` into here like? Does this help performance improvement?
```
if (columnarBatch == null) resultBatch();
```
I am curious how orc reader handle this.
cc @dongjoon-hyun @cloud-fan

Comment (SongYadong) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Thanks for your review.
calling sequence is optional. `initBatch()` was moved to `nextBatch()`, if `hasNext()` called before `initBatch()`, eventually `initBatch()` will be called first of all in `nextBatch()`:
```
  public boolean nextBatch() throws IOException {
     if (columnarBatch == null) initBatch();
```

Orc has a much simple way: 
```
  public boolean nextKeyValue() throws IOException {
    return nextBatch();
  }
```

**Comment by SongYadong** [comment]
Could I ask you to review this PR if you have the time? cc @dongjoon-hyun @cloud-fan

**Comment by dongjoon-hyun** [comment]
@SongYadong . `0.6%` seems to be too small to do this. Do you have another benchmark result to show the benefit more clearly?

**Comment by SongYadong** [comment]
@dongjoon-hyun .  You are right, DataSourceReadBenchmark result show the benefit is too small even in some cases is covered up by fluctuation.

Java HotSpot(TM) 64-Bit Server VM 1.8.0_141-b15 on Windows 7 6.1
Intel64 Family 6 Model 42 Stepping 7, GenuineIntel

Before:
```
Parquet Reader Single TINYINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       330 /  334         47.7          21.0       1.0X
ParquetReader Vectorized -> Row                213 /  301         73.7          13.6       1.5X
```
After:
```
Parquet Reader Single TINYINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       292 /  366         53.8          18.6       1.0X
ParquetReader Vectorized -> Row                254 /  286         62.0          16.1       1.2X
```



Before:
```
Parquet Reader Single SMALLINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       391 /  425         40.2          24.9       1.0X
ParquetReader Vectorized -> Row                371 /  407         42.4          23.6       1.1X
```
After:
```
Parquet Reader Single SMALLINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       435 /  485         36.1          27.7       1.0X
ParquetReader Vectorized -> Row                398 /  440         39.5          25.3       1.1X
```



Before:
```
Parquet Reader Single INT Column Scan:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       453 /  516         34.7          28.8       1.0X
ParquetReader Vectorized -> Row                542 /  563         29.0          34.5       0.8X
```
After:
```
Parquet Reader Single INT Column Scan:   Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       533 /  602         29.5          33.9       1.0X
ParquetReader Vectorized -> Row                549 /  570         28.6          34.9       1.0X
```




Before:
```
Parquet Reader Single BIGINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       800 /  817         19.7          50.9       1.0X
ParquetReader Vectorized -> Row                530 /  686         29.7          33.7       1.5X
```
After:
```
Parquet Reader Single BIGINT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       692 /  847         22.7          44.0       1.0X
ParquetReader Vectorized -> Row                580 /  610         27.1          36.9       1.2X
```




Before:
```
Parquet Reader Single FLOAT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       467 /  543         33.7          29.7       1.0X
ParquetReader Vectorized -> Row                457 /  507         34.4          29.1       1.0X
```
After:
```
Parquet Reader Single FLOAT Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       584 /  600         26.9          37.1       1.0X
ParquetReader Vectorized -> Row                546 /  555         28.8          34.7       1.1X
```




Before:
```
Parquet Reader Single DOUBLE Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       801 /  808         19.6          50.9       1.0X
ParquetReader Vectorized -> Row                590 /  668         26.7          37.5       1.4X
```
After:
```
Parquet Reader Single DOUBLE Column Scan: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       833 /  858         18.9          53.0       1.0X
ParquetReader Vectorized -> Row                672 /  722         23.4          42.7       1.2X
```


Before:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                      2098 / 2263          5.0         200.1      13.3X
```
After:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                      1943 / 2084          5.4         185.3      14.0X
```



Before:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                      1930 / 1980          5.4         184.0      15.0X
```
After:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                      1873 / 1875          5.6         178.6      16.7X
```


Before:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       301 /  356         34.9          28.7      83.1X
```
After:
```
String with Nulls Scan:                  Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
ParquetReader Vectorized                       506 /  542         20.7          48.3      53.0X
```

**Comment by dongjoon-hyun** [comment]
Thank you for experiment, @SongYadong . IMO, if the gain is not obvious, we had better keep the current proven one because new one can introduce some future regressions (or even bugs).

**Comment by SongYadong** [comment]
Sounds reasonable. I will close this PR. Thank you!

**Pull request closed without merging**